File: src/lib/jobs-scraper/service.ts
import "server-only";

import { z } from "zod";
import {
  DEFAULT_TEXT_GENERATION_MODEL_ID,
  getProviderModelId,
} from "@/lib/ai/text-generation/models";
import { generateGoogleJson } from "@/lib/ai/text-generation/providers/google-ai-studio";
import { TEXT_GENERATION_PROVIDER_IDS } from "@/lib/ai/text-generation/types";
import {
  collectJobDocuments,
  type JobDocument,
} from "@/lib/jobs-scraper/crawler";
import { logJobsScraper } from "@/lib/jobs-scraper/logging";
import {
  type DiscoveredJob,
  getJobDateWindow,
  getJobRejectionReasons,
  type JobScraperConfig,
  type JobScraperState,
  jobExtractionSchema,
  type PotentialJob,
  parseJobExtraction,
  type SavedJob,
} from "@/lib/jobs-scraper/schema";
import {
  getJobScraperConfig,
  getJobScraperState,
  persistJobScan,
} from "@/lib/jobs-scraper/storage";

export async function scanJobs(
  scanId: string,
  userId: string,
): Promise<JobScraperState> {
  const scanStartedAt = Date.now();
  const scannedAt = new Date();
  const config = await getJobScraperConfig(userId);
  const dateWindow = getJobDateWindow(config, scannedAt);

  logJobsScraper("info", "scan.started", scanId, {
    dateWindow,
    roles: config.roles,
    sources: config.sources,
    timeRange: config.timeRange,
    workModes: {
      philippines: config.philippinesWorkModes,
      worldwide: config.worldwideWorkModes,
    },
  });

  const { documents, sourceIssues } = await collectJobDocuments(scanId, config);

  logJobsScraper("info", "documents.collected", scanId, {
    documents: documents.map(({ source, url }) => ({ source, url })),
    documentCount: documents.length,
  });

  if (documents.length === 0) {
    logJobsScraper("info", "gemini.skipped", scanId, {
      reason: "no_documents",
    });
    await persistJobScan(userId, scanId, scannedAt, []);
    const state = await getJobScraperState(userId, sourceIssues);

    logCompletedScan(scanId, scanStartedAt, state);
    return state;
  }

  const providerModelId = getProviderModelId({
    modelId: DEFAULT_TEXT_GENERATION_MODEL_ID,
    providerId: TEXT_GENERATION_PROVIDER_IDS.GOOGLE_AI_STUDIO,
  });
  const geminiStartedAt = Date.now();
  const text = await generateGoogleJson({
    providerModelId,
    responseJsonSchema: z.toJSONSchema(jobExtractionSchema),
    prompt: buildExtractionPrompt(scannedAt, config, documents),
  });
  const result = parseJobExtraction(JSON.parse(text) as unknown);

  logJobsScraper("info", "gemini.classification.completed", scanId, {
    durationMs: Date.now() - geminiStartedAt,
    matchCandidates: result.jobs.map(toLoggedJob),
    model: providerModelId,
    potentialCandidates: result.potentialJobs.map(toLoggedJob),
    rejectedCandidates: result.rejectedJobs.map(toLoggedJob),
  });

  const descriptions = new Map(
    documents.map((document) => [document.url.toLowerCase(), document.content]),
  );
  const classified = classifyScannedJobs(result, config, scannedAt).map(
    (entry) => ({
      ...entry,
      description: descriptions.get(entry.job.url.toLowerCase()) ?? "",
    }),
  );

  logJobsScraper("info", "selection.completed", scanId, {
    jobs: classified.map(({ classification, job, reviewReasons }) => ({
      ...toLoggedJob(job),
      classification,
      reviewReasons,
    })),
  });

  await persistJobScan(userId, scanId, scannedAt, classified);

  const state = await getJobScraperState(userId, sourceIssues);
  logJobsScraper("info", "jobs.persisted", scanId, {
    newJobCount: state.newJobCount,
    savedJobCount: state.savedJobCount,
  });
  logCompletedScan(scanId, scanStartedAt, state);
  return state;
}

/**
 * Grades every extracted listing against the configured criteria and settles it
 * into one bucket. A listing Gemini proposed as a match or a potential match is
 * demoted to REJECTED when the deterministic criteria disagree, and it carries
 * the reasons for that verdict so the scan stays reviewable.
 */
export function classifyScannedJobs(
  result: {
    jobs: DiscoveredJob[];
    potentialJobs: PotentialJob[];
    rejectedJobs: PotentialJob[];
  },
  config: JobScraperConfig,
  scannedAt: Date,
): {
  classification: SavedJob["classification"];
  job: DiscoveredJob | PotentialJob;
  reviewReasons: string[];
}[] {
  const candidates = [
    ...result.jobs.map((job) => ({ job, proposed: "MATCH" as const })),
    ...result.potentialJobs.map((job) => ({
      job,
      proposed: "POTENTIAL" as const,
    })),
    ...result.rejectedJobs.map((job) => ({
      job,
      proposed: "REJECTED" as const,
    })),
  ];
  const seen = new Set<string>();

  return candidates.flatMap(({ job, proposed }) => {
    const dedupeKey = job.url.toLowerCase();

    if (seen.has(dedupeKey)) {
      return [];
    }

    seen.add(dedupeKey);
    const failures = getJobRejectionReasons(job, config, scannedAt);
    const classification =
      proposed === "REJECTED" || failures.length > 0 ? "REJECTED" : proposed;
    const modelReasons = "reviewReasons" in job ? job.reviewReasons : [];

    return [
      {
        classification,
        job,
        reviewReasons: [...new Set([...modelReasons, ...failures])].slice(0, 6),
      },
    ];
  });
}

function logCompletedScan(
  scanId: string,
  scanStartedAt: number,
  state: JobScraperState,
) {
  logJobsScraper(
    state.sourceIssues.length ? "warn" : "info",
    "scan.completed",
    scanId,
    {
      classificationCounts: state.classificationCounts,
      durationMs: Date.now() - scanStartedAt,
      newJobCount: state.newJobCount,
      savedJobCount: state.savedJobCount,
      sourceIssues: state.sourceIssues,
    },
  );
}

function toLoggedJob(job: DiscoveredJob | PotentialJob) {
  return {
    company: job.company,
    country: job.country,
    level: job.level,
    location: job.location,
    matchedSkills: job.matchedSkills,
    postedAt: job.postedAt,
    postedDate: job.postedDate,
    source: job.source,
    title: job.title,
    url: job.url,
    workMode: job.workMode,
    ...("reviewReasons" in job ? { reviewReasons: job.reviewReasons } : {}),
  };
}

function buildExtractionPrompt(
  scannedAt: Date,
  config: JobScraperConfig,
  documents: JobDocument[],
): string {
  const dateWindow = getJobDateWindow(config, scannedAt);
  const crawledDocuments = documents
    .map(
      (document, index) =>
        `DOCUMENT ${index + 1}\nSOURCE: ${document.source}\nURL: ${document.url}\n${document.content}`,
    )
    .join("\n\n---\n\n");
  const requiredTechnologyRule =
    config.requiredTechnologies.length > 0
      ? `The listing must mention at least one of: ${config.requiredTechnologies.join(", ")}.`
      : "No technology is required.";
  const dateRule =
    config.timeRange === "LAST_HOUR"
      ? `Include only listings explicitly posted during the last hour relative to ${scannedAt.toISOString()}. Convert an explicit relative time such as "35 minutes ago" to postedAt; if it cannot be established, return the listing as potential.`
      : `Include only listings posted from ${dateWindow.fromDate} through ${dateWindow.toDate}, inclusive.`;

  return `Extract qualifying job listings from the crawled documents below. Do not search the web and do not invent missing details.

User criteria:
- Roles: ${config.roles.join(", ")}.
- Included levels: ${config.includedLevels.join(", ")}.
- Excluded levels: ${config.excludedLevels.join(", ")}.
- ${requiredTechnologyRule}
- Excluded technologies: ${config.excludedTechnologies.join(", ") || "none"}.
- Worldwide work modes: ${config.worldwideWorkModes.join(", ") || "none"}.
- Additional work modes allowed for jobs in the Philippines: ${config.philippinesWorkModes.join(", ") || "none"}.
- ${dateRule}

Rules:
- A title must match one configured role. The level must match an included level and must not match an excluded level.
- Worldwide modes apply to every country. Philippines modes apply only when the job country is the Philippines.
- A listing that mentions a configured excluded technology does not qualify. JavaScript is not Java.
- Use the document's SOURCE and exact URL. LinkedIn URLs must start with https://www.linkedin.com/jobs/view/ and JobStreet URLs must start with https://ph.jobstreet.com/job/.
- Write every field in English. If the document is in another language, translate it, including the country name and the posting label.
- Set postedDate to the Philippine calendar date. Preserve the site's visible posting label in postedText. Set postedAt only when an exact timestamp or an explicit relative posting time can support it.
- Return exactly one entry per document, in exactly one of the three lists, and never repeat a URL. Every document must be accounted for. The summary must state the relevant stack and work arrangement.

Classify each document into one list:
- jobs: it satisfies every criterion above.
- potentialJobs: the role and level qualify, but the date, work mode, or required technology is missing or unclear from the document. Explain each uncertainty in reviewReasons.
- rejectedJobs: it fails at least one criterion, or the document is not a readable job listing. State every disqualifying detail in reviewReasons, most important first, each one specific (name the excluded technology, the seniority wording, the country, or the posting date you read).

Write reviewReasons as short sentences addressed to the person reviewing the scan, so they can judge whether the verdict was correct. Fill in as much of the listing as the document supports even when rejecting it; use "Unclear" for an unknown work mode and null for an unknown posting date.

SCAN TIME: ${scannedAt.toISOString()}

CRAWLED DOCUMENTS:
${crawledDocuments}`;
}

File: apps/ai/app/modules/jobs_scraper/service.py
"""Crawl, extract with Gemini, then grade. No persistence lives here.

Port of apps/web/src/lib/jobs-scraper/service.ts with every storage call
removed: the caller (Nest) owns the config it passes in and the rows it writes
out of the result.
"""

from __future__ import annotations

import time
from datetime import datetime

from app.core.config import get_settings
from app.core.logging import elapsed_ms, log_jobs_scraper
from app.lib.jobs_scraper.rules import (
    get_job_date_window,
    get_job_rejection_reasons,
    get_job_source_from_url,
    get_source_job_id,
)
from app.lib.llm.gemini import extract_jobs
from app.lib.scraping.crawl4ai import collect_job_documents
from app.modules.jobs_scraper.schemas import (
    DiscoveredJob,
    JobClassification,
    JobDocument,
    JobExtraction,
    JobScraperConfig,
    JobSourceIssue,
    PotentialJob,
    ScannedJob,
    ScanResult,
)


async def scan_jobs(
    scan_id: str, config: JobScraperConfig, scanned_at: datetime
) -> ScanResult:
    scan_started_at = time.monotonic()
    from_date, to_date = get_job_date_window(config, scanned_at)

    log_jobs_scraper(
        "info",
        "scan.started",
        scan_id,
        {
            "dateWindow": {"fromDate": from_date, "toDate": to_date},
            "roles": config.roles,
            "sources": config.sources,
            "timeRange": config.time_range,
            "workModes": {
                "philippines": config.philippines_work_modes,
                "worldwide": config.worldwide_work_modes,
            },
        },
    )

    documents, source_issues = await collect_job_documents(scan_id, config)

    log_jobs_scraper(
        "info",
        "documents.collected",
        scan_id,
        {
            "documents": [
                {"source": document.source, "url": document.url}
                for document in documents
            ],
            "documentCount": len(documents),
        },
    )

    if not documents:
        log_jobs_scraper("info", "gemini.skipped", scan_id, {"reason": "no_documents"})
        return _complete(
            scan_id, scanned_at, scan_started_at, [], source_issues, documents
        )

    settings = get_settings()
    gemini_started_at = time.monotonic()
    result = await extract_jobs(scanned_at, config, documents)

    log_jobs_scraper(
        "info",
        "gemini.classification.completed",
        scan_id,
        {
            "durationMs": elapsed_ms(gemini_started_at),
            "matchCandidates": [_to_logged_job(job) for job in result.jobs],
            "model": settings.gemini_model,
            "potentialCandidates": [
                _to_logged_job(job) for job in result.potential_jobs
            ],
            "rejectedCandidates": [_to_logged_job(job) for job in result.rejected_jobs],
        },
    )

    descriptions = {
        document.url.lower(): document.content for document in documents
    }
    classified = [
        _to_scanned_job(entry, descriptions)
        for entry in classify_scanned_jobs(result, config, scanned_at)
    ]

    log_jobs_scraper(
        "info",
        "selection.completed",
        scan_id,
        {
            "jobs": [
                {
                    "classification": job.classification,
                    "reviewReasons": job.review_reasons,
                    "source": job.source,
                    "sourceJobId": job.source_job_id,
                    "title": job.title,
                    "url": job.url,
                }
                for job in classified
            ]
        },
    )

    return _complete(
        scan_id, scanned_at, scan_started_at, classified, source_issues, documents
    )


def classify_scanned_jobs(
    result: JobExtraction,
    config: JobScraperConfig,
    scanned_at: datetime,
) -> list[tuple[JobClassification, DiscoveredJob | PotentialJob, list[str]]]:
    """Grades every extracted listing against the configured criteria and
    settles it into one bucket. A listing Gemini proposed as a match or a
    potential match is demoted to REJECTED when the deterministic criteria
    disagree, and it carries the reasons for that verdict so the scan stays
    reviewable.
    """
    candidates: list[tuple[DiscoveredJob | PotentialJob, JobClassification]] = [
        *((job, "MATCH") for job in result.jobs),
        *((job, "POTENTIAL") for job in result.potential_jobs),
        *((job, "REJECTED") for job in result.rejected_jobs),
    ]
    seen: set[str] = set()
    classified: list[
        tuple[JobClassification, DiscoveredJob | PotentialJob, list[str]]
    ] = []

    for job, proposed in candidates:
        dedupe_key = job.url.lower()

        if dedupe_key in seen:
            continue

        seen.add(dedupe_key)
        failures = get_job_rejection_reasons(job, config, scanned_at)
        classification: JobClassification = (
            "REJECTED" if proposed == "REJECTED" or failures else proposed
        )
        model_reasons = (
            job.review_reasons if isinstance(job, PotentialJob) else []
        )
        reasons = list(dict.fromkeys([*model_reasons, *failures]))[:6]
        classified.append((classification, job, reasons))

    return classified


def _to_scanned_job(
    entry: tuple[JobClassification, DiscoveredJob | PotentialJob, list[str]],
    descriptions: dict[str, str],
) -> ScannedJob:
    classification, job, review_reasons = entry
    # A listing rejected for a bad URL can name a source its URL does not
    # belong to; fall back to the model's own source so the id stays derivable.
    source = get_job_source_from_url(job.url) or job.source

    return ScannedJob(
        classification=classification,
        source=source,
        source_job_id=get_source_job_id(job.url, source),
        url=job.url,
        title=job.title,
        company=job.company,
        location=job.location,
        country=job.country,
        work_mode=job.work_mode,
        level=job.level,
        posted_date=job.posted_date,
        posted_at=job.posted_at,
        posted_text=job.posted_text,
        summary=job.summary,
        description=descriptions.get(job.url.lower(), ""),
        matched_skills=job.matched_skills,
        review_reasons=review_reasons,
    )


def _complete(
    scan_id: str,
    scanned_at: datetime,
    scan_started_at: float,
    jobs: list[ScannedJob],
    source_issues: list[JobSourceIssue],
    documents: list[JobDocument],
) -> ScanResult:
    counts: dict[str, int] = {"MATCH": 0, "POTENTIAL": 0, "REJECTED": 0}

    for job in jobs:
        counts[job.classification] += 1

    log_jobs_scraper(
        "warn" if source_issues else "info",
        "scan.completed",
        scan_id,
        {
            "classificationCounts": counts,
            "documentCount": len(documents),
            "durationMs": elapsed_ms(scan_started_at),
            "sourceIssues": [
                issue.model_dump(by_alias=True) for issue in source_issues
            ],
        },
    )

    return ScanResult(
        scan_id=scan_id,
        scanned_at=scanned_at,
        document_count=len(documents),
        jobs=jobs,
        source_issues=source_issues,
    )


def _to_logged_job(job: DiscoveredJob | PotentialJob) -> dict[str, object]:
    logged: dict[str, object] = {
        "company": job.company,
        "country": job.country,
        "level": job.level,
        "location": job.location,
        "matchedSkills": job.matched_skills,
        "postedAt": job.posted_at,
        "postedDate": job.posted_date,
        "source": job.source,
        "title": job.title,
        "url": job.url,
        "workMode": job.work_mode,
    }

    if isinstance(job, PotentialJob):
        logged["reviewReasons"] = job.review_reasons

    return logged

